fix #7269: complete support vscode.workspace.fs API

akosyakov opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->

This PR replaces our filesystem implementation with VS Code one to provide first class support for different fs providers and features like dealing with encoding, cross provider operations, support working copy modifications and so on.

- [x] porting VS Code implementation
  - [x] FileSystemService
     - [x] activation event
  - [x] DiskFileSystemProvider
  - [x] RemoteFileSystemProvider
  - [x] WorkingCopyFileService
  - [x] TextFileService
  - [x] EncodingOracle
    - [x] support encoding overrides
  - [x] file a CQ: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=22318
  - [x] add copyright comments
  - [x] add links to copied code as comments
- [x] adjusting clients
  - [x] Extension Host Process
  - [x] FileTree, FileTreeModel
  - [x] FileTreeDialog
  - [x] Explorer
  - [x] FileSystemFrontendContribution
  - [x] Resource
  - [x] MonacoTextModel
  - [x] DebugConfigurationManager
  - [x] FileDialogService
  - [x] FileDownloadHandler
  - [x] GitDiffContribution
  - [x] MonacoWorkspace
  - [x] NavigatorDiff
  - [x] DialogsMainImpl
  - [x] ScmHistoryWidget
  - [x] SearchInWorkspaceFrontendContribution
  - [x] TaskConfigurationManager
  - [x] TerminalFrontendContribution
  - [x] TerminalLinkmatcherFiles
  - [x] DiffService
  - [x] QuickOpenWorkspace
  - [x] WorkspaceCommands
  - [x] WorkspaceFrontendContribution
  - [x] WorkspaceService
  - [x] GitQuickOpenService
  - [x] MiniBrowserContent
  - [x] MiniBrowserEndpoint
  - [x] HostedPluginSupport (browser) and PluginPathsServiceImpl (node)
  - [x] PreferencesContribution
  - [x] MonacoSnippetSuggestProvider
  - [x] MonacoThemingService
  - [x] PluginIconThemeService
  - [x] GitScmProvider
  - [x] WorkspaceDeleteHandler
  - [x] DebugSession
  - [x] ScmQuickOpenService
  - [x] WorkspaceVariableContribution
  - [x] GitRepositoryProvider
  - [x] MarkerManager
- [x] deprecating Theia services
  - [x] FileSystem
    - [x] move `getCurrentUserHome` to `EnvVariableServer`
    - [x] move `getDrives` to `EnvVariableServer`
  - [x] FileSystemWatcher
  - [x] replace UserStorageService with userdata fs provider
    - [x] hook `vscode-userdata` and `vscode-remote` fs providers delegating to `userstorage` and `file` schemes
    - [x] make sure that user storage paths are absolute
  - [x] remove FileSystemNode, FileSystem should NOT be bound on the backend since it reflects user triggered operations
    - [x] what about `FileSystemNodeOptions`? is it not used in practice and there are user preferences to control trash and encoding
  - [x] stub FileSystem on frontend with `FileService`
  - [x] stub FileSystemWatcher with `FileService`
    - [x] what about FileShouldOverwrite? review create/write places and apply overwrite logic
  - [x] document in breaking changes
- [x] preferences
  - [x] files.watcherExclude
  - [x] files.enableTrash
    - [x] `DiskFileSystemProvider` should support `useTrash` then
  - [x] files.encoding
  - [x] files.autoGuessEncoding
  - [x] files.participants.timeout
- [x] fix tests
- [x] fix vscode api tests

Out of scope:
- fixing URI comparisons, see https://nodejs.org/en/docs/guides/working-with-different-filesystems/ that toString and toLowerCase is not good
- complete untitled handling
- custom unix/windows file watching
- handling all configuration scopes: https://github.com/microsoft/vscode/blob/45aa0bf525b08d7e23c9b10ca65f1de55949e601/src/vs/platform/configuration/common/configurationRegistry.ts#L85-L110
- support of override readonly on write
- support of write elevated
- support text only on read, i.e. fail on binary
- disabling modification in readonly resources, i.e. changing content, rename, delete, create file/folders within and so on

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->

- file dialogs:
  - [x] open and save dialogs should still work
  - [x] open and save dialogs should work when triggered from VS Code extensions
  - [x] displaying drivers should work
- file encoding:
  - [x] reopen with encoding should work
  - [x] save with encoding should work, after save reopening the editor should detect the proper encoding
  - [x] changing the default encoding should work, but utf8 should be used for setting files always
- dirty state management
  - [x] on file delete dirty editors should be marked as deleted
  - [x] on file delete not dirty editors should be closed
  - [x] on file add a deleted marker should be removed from dirty editors
  - [x] on file move by a user editors should be moved
  - [x] on file move by a system or api editors should be marked as deleted
- navigator
  - [x] restoring nodes of old format should work
  - [x] copy/move should work
  - [x] moving files between folders is reflected properly
  - [x] disk file changes are reflected properly 
  - [x] `Select for Compare` and `Compare with Selected` commands should work
  - [x] should be possible to download single file
  - [x] should be possible to download multiple files
- workspace management
  - [x] open file and open with...
  - [x] open folder
  - [x] open workspace
  - [x] close workspace
  - [x] open recent workspace
  - [x] new file/folder
  - [x] save file
  - [x] save file as...
  - [x] file(s) rename
  - [x] file(s) delete
  - [x] file(s) duplicate
  - [x] compare with each other
  - [x] add/remove folders to/from a workspace
  - [x] save a workspace file
  - [x] workspace roots should be watched
  - [x] files should be excluded according to `files.watcherExclude` from watching
- [x] `Git Diff: Compare with...` command should work
- [x] keeping fs watching after reconnection
- [x] fs changes from providers (disk)
  - [x] add
  - [x] update
  - [x] delete
- [x] fs changes from users
  - [x] create
  - [x] move
  - [x] delete
  - [x] copy
- text editors
  - [x] read/write content (editors, preferences, search in workspace replace preview for a node)
  - [x] no performance degradation on save (large files, incremental updates)
  - [x] bulk resource edits should be applied, i.e rename ts class
  - [x] out of sync
- preferences
  - [x] creating new launch.json
  - [x] opening and creating tasks.json file should work
  - [x] opening settings files from the preference view should work
  - [x] opening new JSON file from the preference view should work
- source control
  - [x] open the history for files from the navigator
  - [x] `Git: Clone` command should work
  - [x] `Git: Discard All Changes` command should work
  - [x] `SCM: Change Repository...` command should work
  - [x] git repositories should be refreshed on changes in the workspace
- search in workspace
  - [x] `Find in Folder` command should work from the navigator
- terminal
  - [x] open in terminal should work from the navigator for files and folders
  - [x] directory URIs should not match, but file URIs should
- preview (mini browser)
  - [x] refreshing on file changes should work
  - [x] fetching various files should work, i.e. html, images, svg and so on
- vscode extensions
  - [x] storing global and workspace data should work
  - [x] restoring existing global and workspace data should work
  - [x] snippets should be properly loaded from the disk
  - [x] themes should be properly loaded from the disk
  - [x] icon themes should be properly loaded from the disk and applied for the file and 
    - [x] watchable icon themes are reloaded on the disk changes
  - [x] python (disk fs access) - https://github.com/microsoft/vscode-python/releases/tag/2020.5.86806 (last version which does not require custom editors API)
  - [x] typescript (fs events from users)
  - [x] java (fs events from users)
- task/debug
  - [x] navigation to the loaded node modules should work from the stack view
  - [x] `cwd`, `workspaceFolder`, `workspaceRoot`, and `file` variables should be resolved properly
- problems
  - [x] markers should be removed if a file or a parent folder is deleted

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

akosyakov commented:
@lmcbout Could you test please what is the experience of editing large files for you? We have to drop incremental update optimisation, since new FS API is based on binary streams, not strings.

akosyakov commented:
@marcdumais-work @svenefftinge I'm going to start documenting breaking changes and filing a CQ for ported code tomorrow. Rather than that I've tested everything as good as I could. It should be ready for review and testing.

akosyakov commented on configs/base.eslintrc.json:
```diff
@@ -21,5 +22,8 @@
-}
+  ],
+  "rules": {
+    "deprecation/deprecation": "off"
```
figure out why deprecated APIs are not marked as such in the editor

akosyakov replied:
now enabled only as warnings

akosyakov commented on package.json:
```diff
@@ -7,24 +7,25 @@
   },
   "resolutions": {
-    "**/@types/node": "~10.3.6",
+    "**/@types/node": "~10.12.0",
```
try to keep old node version if getting Node.js 12 is not possible next week

akosyakov replied:
cc @marcdumais-work there are different deps update, but i think we are better postpone licenses check before we settle on the final version of the PR

akosyakov replied:
I reverted it to current Node version

akosyakov commented on packages/core/src/common/application-error.ts:
```diff
@@ -32,11 +32,7 @@ export namespace ApplicationError {
     }
-    const codes: number[] = [];
     export function declare<C extends number, D>(code: C, factory: (...args: any[]) => Literal<D>): Constructor<C, D> {
-        if (codes.indexOf(code) !== -1) {
```
Not sure about it. It makes defining error codes per a service easier, since you don't need to care about errors from other services. But if a client using several services it could lead to issues.

akosyakov replied:
reverted

akosyakov commented on packages/filesystem/src/common/remote-file-system-provider.ts:
```diff
@@ -0,0 +1,390 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+/*---------------------------------------------------------------------------------------------
```
remove, there is nothing copied

akosyakov commented:
VS Code woskpaces.fs api tests are passing:
<img width="1904" alt="Screenshot 2020-06-28 at 13 20 56" src="https://user-images.githubusercontent.com/3082655/85954642-4bac8400-b979-11ea-981b-90fe927ae161.png">
VS Code woskpaces events api tests are passing:
<img width="1904" alt="Screenshot 2020-06-28 at 16 09 04" src="https://user-images.githubusercontent.com/3082655/85954646-50713800-b979-11ea-8190-4bd3054fcd11.png">

akosyakov commented:
a CQ: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=22318

akosyakov commented:
I've updated `CHANGELOG`. I will do some measurements with big files now like compiler.ts from Typescript repo and a file provided by @marcdumais-work

akosyakov commented:
With checker.ts each save transfer around 400kb with gzipping compare to several bytes as before. From user perspective I don't really feel the difference than latency is low, but as soon as it get higher the difference become noticeable. I am not sure how we should approach it. We could either go as is and look at the feedback or try to workaround the whole buffer based abstraction for disk implementation of the filesystem as an optimisation. The latter is the leaking abstraction.

I've tested as well with VS Codespaces and they transferred the whole content as well. The difference only that they stream it. It allows to open bigger files (https://code.visualstudio.com/blogs/2018/03/23/text-buffer-reimplementation#_avoid-the-string-concatenation-trap) but does not make much difference in regards to latency. We could enable buffering later as well if someone wants to open very big file in Electron for instance.

akosyakov commented on examples/api-tests/src/typescript.spec.js:
```diff
@@ -260,7 +257,9 @@ module.exports = (port, host, argv) => Promise.resolve()
 
-    it('document formating should be visible and enabled', async () => {
+    it('document formating should be visible and enabled', async function () {
+        this.timeout(0);
```
left over!

lmcbout commented:
Start testing this PR.
 Testing env:
- UBUNTU 18.04
- Chrome

1-  Opening a file generates the following:
root ERROR [hosted-plugin: 31783] Promise rejection not handled in one second: TypeError: r.SemanticTokensLegend is not a constructor , reason: TypeError: r.SemanticTokensLegend is not a constructor
root ERROR [hosted-plugin: 31783] With stack trace: TypeError: r.SemanticTokensLegend is not a constructor
    at a.getLegend (/home/lmcbout/theia/theia/plugins/vscode-builtin-typescript-language-features/extension/dist/extension.js:1:182974)

2- Scenario:
    - Put the Auto Save OFF
    - Open a file (from Theia: application-package-manager.ts )
    - Do any modification in the file
    - Manually save the file (CTRL + S)

SERVER SIDE
--------------------
root ERROR onWillSave listeners should provide edits, not directly alter the document.

Client side:
---------------
logger-protocol.ts:112 root ERROR onWillSave listeners should provide edits, not directly alter the document.

Note: This occurs on the first occurence, sometimes on the second save as well. Do not happen on the Theia:latest commit


3-  Another error reported. I don't know if this is related to this PR:
CLIENT SIDE  
------------------
CellMeasurer.js:102 GET http://localhost:3000/hostedPlugin/vscode_npm/%24%28run%29 404 (Not Found)
_getCellMeasurements @ CellMeasurer.js:102
_maybeMeasureCell @ CellMeasurer.js:131
componentDidUpdate @ CellMeasurer.js:62
commitLifeCycles @ react-dom.development.js:19835
commitLayoutEffects @ react-dom.development.js:22803
callCallback @ react-dom.development.js:188
invokeGuardedCallbackDev @ react-dom.development.js:237
invokeGuardedCallback @ react-dom.development.js:292
commitRootImpl @ react-dom.development.js:22541
unstable_runWithPriority @ scheduler.development.js:653
runWithPriority$1 @ react-dom.development.js:11039
commitRoot @ react-dom.development.js:22381
finishSyncRender @ react-dom.development.js:21807
performSyncWorkOnRoot @ react-dom.development.js:21793
(anonymous) @ react-dom.development.js:11089
unstable_runWithPriority @ scheduler.development.js:653
runWithPriority$1 @ react-dom.development.js:11039
flushSyncCallbackQueueImpl @ react-dom.development.js:11084
flushSyncCallbackQueue @ react-dom.development.js:11072
scheduleUpdateOnFiber @ react-dom.development.js:21199
enqueueForceUpdate @ react-dom.development.js:12678
../../node_modules/react/cjs/react.development.js.Component.forceUpdate @ react.development.js:490
recomputeGridSize @ Grid.js:355
recomputeRowHeights @ List.js:176
../../packages/core/lib/browser/tree/tree-widget.js.TreeWidget.forceUpdate @ tree-widget.tsx:339
../../packages/core/lib/browser/tree/tree-widget.js.TreeWidget.updateScrollToRow @ tree-widget.tsx:303
../../packages/core/lib/browser/tree/tree-widget.js.TreeWidget.doUpdateRows @ tree-widget.tsx:289
....

vince-fugnitto commented:
@lmcbout 

> 1- Opening a file generates the following:
> root ERROR [hosted-plugin: 31783] Promise rejection not handled in one second: TypeError: r.SemanticTokensLegend is not a constructor , reason: TypeError: r.SemanticTokensLegend is not a constructor
> root ERROR [hosted-plugin: 31783] With stack trace: TypeError: r.SemanticTokensLegend is not a constructor
> at a.getLegend (/home/lmcbout/theia/theia/plugins/vscode-builtin-typescript-language-features/extension/dist/extension.js:1:182974)

This error is on `master`, we do not yet have plugin support for `r.SemanticTokensLegend`.

> 
> 2- Scenario:
> - Put the Auto Save OFF
> - Open a file (from Theia: application-package-manager.ts )
> - Do any modification in the file
> - Manually save the file (CTRL + S)
> ## SERVER SIDE
> root ERROR onWillSave listeners should provide edits, not directly alter the document.
> 
> ## Client side:
> logger-protocol.ts:112 root ERROR onWillSave listeners should provide edits, not directly alter the document.

This error is on `master`, and was introduced in your previous pull-request: https://github.com/eclipse-theia/theia/pull/4012
- https://github.com/eclipse-theia/theia/pull/4012#issuecomment-455315152

lmcbout commented:
Testing performance with auto-save ON:
File 100K: OK
File 500K:  Ok, little latency before saving
File 1 Mg:  takes ~ 1 sec after you finish typing. If slow typing, then It waits ~ 1 sec after the last char is in, so OK
File 4 Mg:  takes ~ 2 sec after you finish typing before the auto-save is triggered. Seems OK to me

marcdumais-work commented:
> Testing performance with auto-save ON:
> File 100K: OK
> File 500K: Ok, little latency before saving
> File 1 Mg: takes ~ 1 sec after you finish typing. If slow typing, then It waits ~ 1 sec after the last char is in, so OK
> File 4 Mg: takes ~ 2 sec after you finish typing before the auto-save is triggered. Seems OK to me

Just by this it's a bit hard to understand exactly what's happening and whether it's acceptable or not. Can you provide a recording for the case with 4 Meg file, so we can see for ourselves? The delay is not necessarily the only factor - does it get in the way of the developer?

akosyakov commented:
It is important to measure with different network conditions as well. On a connection with low latency you won't notice how fast increased amount of content is sent, but on a connection with higher latency it would become noticeable. ...I tend to reintroduce the optimisation for `file` protocol. I also noticed that we unnecessary refetch content again on a change with this PR, trying to figure out what is causing it.

lmcbout commented:
Recording with a 4 Mg file with slow typing. Notice the save  (Icon on the file title) after finishing typying.
Auto-Save is not done between each char, but after you stop typing.

![WebviewWith4Mg](https://user-images.githubusercontent.com/419051/86611691-04c41d00-bf7d-11ea-9dd8-510ab36accb1.gif)

marcdumais-work commented:
Thanks for the video Jacques. 

I think you illustrate well the time it takes to save this big file. But I wonder if an average user would be so disciplined to stop typing until the file is saved? i.e. What happens if, while saving, you make another change and save it too?

akosyakov commented:
@marcdumais-work @lmcbout this PR with enabled optimization: https://github.com/eclipse-theia/theia/pull/8143

akosyakov commented:
I've reintroduce the optimisation. In the previous version following was happening on the save:
- the entire content was saved
- but first it was fetched to check for BOM if encoding is UTF8
- afterwards it was fetched again since there was a bug in sync check on frontend

New version:
- send only changes when it is appropriate (as before) for file scheme
- does BOM check on backend for file scheme, for other schemes it only streams necessary amount of bytes
- in sync was fixed before, so content won't be fetched again after save

What is importat for measurement:
- don't look at UI, there are many things happening before save, like waiting for formatting participant and it distorts measurement
- use ws frame inspector to see what kind of communication happening when you save
- use system network conditioner to proxy to simulate latency for ws connection (browsers cannot do it, only for http)

cc @marcdumais-work @lmcbout 

btw issues from https://github.com/eclipse-theia/theia/pull/7908#issuecomment-654279126 are present on master

akosyakov commented:
@marcdumais-work Could you run please license check for runtime dependencies? I'm pretty sure that they are good, but double checking would be better. 🙏

marcdumais-work commented:
> @marcdumais-work Could you run please license check for runtime dependencies? I'm pretty sure that they are good, but double checking would be better.

I just completed the check - 3PP production dependencies look good, license-wise

lmcbout commented:
Testing performance
Using a 4Mg file to test
use ws frame inspector in Chrome

I noticed there is only the difference that is being send and not the complete file, so performance is very good with auto-save on. 
Great job @akosyakov

akosyakov commented:
@marcdumais-work Is it mandatory for us to wait till CQs where we know that code is under MIT to be approved by the foundation? I remember we discussed that we could merge it and do occasional checks of the whole VS Code? It takes too long and no response from them. 

btw it looks like there are already solutions in the wild based on this PR: https://twitter.com/plotlygraphs/status/1281672032955047938

marcdumais-work commented:
> Is it mandatory for us to wait till CQs where we know that code is under MIT to be approved by the foundation? 

Yes, I think we need to wait for CQs about copied code, unless given permission to merge while they finish their investigation (so-called "parallel IP*"). The main vscode repo's license is indeed MIT, but there are bits that are under different licenses. So depending what exactly we copy, it could have an impact. __I've pinged the PMC on the CQ, asking to be granted parallel IP.__ 

> I remember we discussed that we could merge it and do occasional checks of the whole VS Code? It takes too long and no response from them

Yes, that was a proposal by the foundation, for them to deep-analyse whole versions of vscode, making them pre-approved, for us to use/fork (in theory). I think there was an assumption that this would not need to happen too often (i.e. not every vscode release) and that we would/could use these few pre-approved versions most of the time. So maybe more a way to save work for the Foundation, rather than make it more efficient for us :)

In this PR it looks like you copy from the as-of-yet unreleased (at CQ creation time) `1.47.0` version, so it probably would not have helped.


(*): from the Eclipse Project Handbook:
```
What is Parallel IP?

The parallel IP Process allows an Eclipse projects to make use of project code 
contributions and third-party content before they are fully approved by the Eclipse
IP Team. In versions of the IP Policy prior to 2019, projects needed to meet 
specific requirements to be able to leverage the Parallel IP. This is no longer the 
case: full vetting is now always applied in parallel in all cases.
```

tsmaeder commented on packages/plugin-ext/src/plugin/file-system-ext-impl.ts:
```diff
@@ -0,0 +1,396 @@
+
+    registerFileSystemProvider(scheme: string, provider: vscode.FileSystemProvider, options: { isCaseSensitive?: boolean, isReadonly?: boolean } = {}) {
+
+        if (this._usedSchemes.has(scheme)) {
```
I'm not sure this will work in the case of Che: since we're running plugin API in multiple hosts for the same front end, we'll end up running multiple instances of FileSystemExtImpl. Doing the uniqueness check on the ext-side will not work in this case.
@benoitf what do you think?

tsmaeder replied:
Ok, seems that the FileService on the browser side checks for uniqueness on the schemes as well. So it's just an early exit, so to speak. Still it leads to two different error cases for the same error, so I would simply remove this check and rely on the one in the browser.

akosyakov replied:
This code is copied from VS Code repo without changes to align with VS Code extension expectations.

tsmaeder commented on packages/plugin/src/theia.d.ts:
```diff
@@ -4355,512 +4520,380 @@ declare module '@theia/plugin' {
-     */
+	/**
+	 * Enumeration of file types. The types `File` and `Directory` can also be
+	 * a symbolic links, in that case use `FileType.File | FileType.SymbolicLink` and
```
"...can also be a symbolic _link_..." while we're at it.

akosyakov replied:
fixed

akosyakov commented:
@tsmaeder thank your for the review! I will work through comments next week.

CQ was approved btw, so maybe we can land it end next week.

akosyakov commented on packages/core/src/common/event.ts:
```diff
@@ -339,4 +344,65 @@ export namespace WaitUntilEvent {
+        interface AsyncEmitter<T> extends Emitter<T> {
+            deliveryQueue?: Promise<void>
+        }
+        const asyncEmitter: AsyncEmitter<T> = emitter;
```
I wanted to keep `WaitUntilEvent.fire(async)` as a public API, but nvm changed it to `AsyncEmitter.fire` and added docs to `WailUntilEvent.fire` to use `AsyncEmitter` then listeners should be fired async one after another.

akosyakov commented on packages/core/src/common/preferences/preference-schema.ts:
```diff
@@ -81,7 +81,7 @@ export interface PreferenceItem {
     description?: string;
     markdownDescription?: string;
-    scope?: 'application' | 'window' | 'resource' | PreferenceScope;
+    scope?: 'application' | 'machine' | 'window' | 'resource' | 'language-overridable' | 'machine-overridable' | PreferenceScope;
```
yep was going to file follow-up issues as soon as PR gets merged, please see out-of-scope section in the description

akosyakov commented on packages/core/src/common/uri.ts:
```diff
@@ -199,8 +199,17 @@ export default class URI {
+        let left = this.path;
+        let right = uri.path;
+        if (!caseSensitive) {
+            left = new Path(left.toString().toLowerCase());
```
It is the follow up issue too. We are generally bad at it and use `toLowerCase` everywhere. The fix will require not only fixing this method, but fixing all other occurrences to use this method.

btw the good write up on OS incompatibilities (including case sensitivity) while working with Node.js: https://nodejs.org/en/docs/guides/working-with-different-filesystems/

akosyakov commented on packages/core/src/node/env-variables/env-variables-server.ts:
```diff
@@ -57,4 +59,36 @@ export class EnvVariablesServerImpl implements EnvVariablesServer {
+    /**
+     * Filters hidden and system partitions.
+     */
+    protected filterMountpointPath(path: string): boolean {
```
it was move method refactoring as is. I've renamed it anyway, since proposed name reflect intent better.

akosyakov commented on packages/debug/src/browser/debug-configuration-manager.ts:
```diff
@@ -273,17 +273,7 @@ export class DebugConfigurationManager {
-        try {
-            await this.filesystem.setContent(fileStat, content);
-        } catch (e) {
-            if (!FileSystemError.FileExists.is(e)) {
```
This catch should have been removed when we switched from `createFile` to `setContent`. Neither `FileSystem.setContent` or `FileService.write` throw such errors, all other errors are alway propagated.

akosyakov commented on packages/debug/src/browser/debug-session.tsx:
```diff
@@ -153,9 +153,11 @@ export class DebugSession implements CompositeTreeElement {
+        let path;
         if (uri.scheme === 'file') {
-            path = await this.fileSystem.getFsPath(path);
+            path = await this.fileService.fsPath(uri);
```
That's only for `file` scheme which is not delegating.

akosyakov replied:
Ok, I see your point and added `FileService.toUnderlyingResource`.

akosyakov commented on packages/filesystem/src/browser/file-service.ts:
```diff
@@ -0,0 +1,1505 @@
+        return { ...content, encoding, value };
+    }
+
+    async update(resource: URI, changes: TextDocumentContentChangeEvent[], options: UpdateTextFileOptions): Promise<FileStatWithMetadata & { encoding: string }> {
```
It is indeed breaking abstraction in order to optimize some cases for `file` scheme: https://github.com/eclipse-theia/theia/pull/7908#issuecomment-654835676 In the long run I would like to get rid of it.

I would be glad to get rid of encoding too. But how would we apply encoded changes on the backend? Currently we are using such API: https://github.com/eclipse-theia/theia/blob/11ac08c29aadeac697b03497f77ec0405ed249b2/packages/filesystem/src/node/disk-file-system-provider.ts#L806

tsmaeder replied:
I see: encoding the inserts is not really a problem, but transforming the change locations to the encoded version is not trivial.
Still I am confident we could compute the encoded locations by traversing the document and transforming the changes as we go: keeping a single offset for the locations should be enough.
But I would leave that to a separate task.

akosyakov commented on packages/filesystem/src/browser/file-service.ts:
```diff
@@ -0,0 +1,1505 @@
+        }
+
+        // file cannot be directory
+        if ((stat.type & FileType.Directory) !== 0) {
```
not sure I understood, here we check that a file is not directory before writing to it and throw an exception if so, it was the same with our file system before and it is what VS Code expects.

One has to delete a directory and then can write a file.

akosyakov commented on packages/filesystem/src/browser/file-tree/file-tree-widget.tsx:
```diff
@@ -243,4 +245,50 @@ export class FileTreeWidget extends TreeWidget {
 
+    protected deflateForStorage(node: TreeNode): object {
+        const deflated = super.deflateForStorage(node);
+        if (FileStatNode.is(node) && FileStatNodeData.is(deflated)) {
```
It won't compile if one removes the check. `FileStatNodeData` is deflated tree node data with some optional fields which should be deflated.

akosyakov commented on packages/filesystem/src/browser/filesystem-frontend-contribution.ts:
```diff
@@ -70,11 +69,37 @@ export class FileSystemFrontendContribution implements FrontendApplicationContri
-        this.fileSystemWatcher.onDidMove(event => event.waitUntil(this.runEach((uri, widget) => this.applyMove(uri, widget, event))));
+        this.fileService.onDidFilesChange(event => this.run(() => this.updateWidgets(event)));
+        this.fileService.onWillRunUserOperation(event => {
+            this.queueUserOperation(event);
```
With new file service timing of external and internal FS change events changed which requires explicit ordering of external events to be processed before internal events. Before with some lucky timing it was the case always.

> Before, we only did the "commit" after success. What's the reason for changing this? Was there a bug?

I am not sure what you mean. The code before will "commit" on success and "revert" on failure. Code right now is doing the same.

akosyakov replied:
That's another follow-up btw: https://github.com/eclipse-theia/theia/issues/8138I will reconsider this code later this week but with another PR. External and internal events should be handled differently.

akosyakov commented on packages/filesystem/src/browser/filesystem-frontend-module.ts:
```diff
@@ -18,45 +18,177 @@ import '../../src/browser/style/index.css';
+                    const stat = await fileService.resolve(new URI(uri), { resolveMetadata: true });
+                    return convertStat(stat);
+                } catch (e) {
+                    return undefined;
```
good catch, thank you

akosyakov commented on packages/filesystem/src/browser/filesystem-frontend-module.ts:
```diff
@@ -18,45 +18,177 @@ import '../../src/browser/style/index.css';
+                encoding?: string | undefined;
+                overwriteEncoding?: string | undefined;
+            } | undefined): Promise<FileStat> {
+                throw new Error('Method not implemented.');
```
Implemented how. In first iteration i did not want to introduce `update` method, but perf testing showed significant regressions, so it was added

tsmaeder replied:
Implemented _now_ ;-)

akosyakov commented on packages/filesystem/src/browser/filesystem-frontend-module.ts:
```diff
@@ -18,45 +18,177 @@ import '../../src/browser/style/index.css';
+                }
+            }
+            async getEncoding(uri: string): Promise<string> {
+                const { encoding } = await fileService.read(new URI(uri));
```
- I don't expect this method to be used by any clients. 
- To detect encoding one has to read to be able to guess. In https://github.com/eclipse-theia/theia/pull/8152 this method is replaced with stream based reading, so unnecessary data won't be fetched anyway.

akosyakov commented on packages/filesystem/src/common/delegating-file-system-provider.ts:
```diff
@@ -0,0 +1,166 @@
+} from './files';
+import type { TextDocumentContentChangeEvent } from 'vscode-languageserver-protocol';
+
+export interface DelegatingFileSystemProviderOptions {
```
- `UriTransformer` does not indicate that it is only for `DelegatingFileSystemProvider`
- `DelegatingFileSystemProviderOptions` could be extended to pass additional options.

I've refactored it to and added js-doc:
```ts
export namespace DelegatingFileSystemProvider {
    export interface Options {
        uriConverter: URIConverter
    }
    export interface URIConverter {
        /**
         * Converts to an underlying fs provider resource format.
         */
        to(resource: URI): URI;
        /**
         * Converts from an underlying fs provider resource format.
         */
        from(resource: URI): URI | undefined;
    }
}
```

akosyakov commented on packages/filesystem/src/common/delegating-file-system-provider.ts:
```diff
@@ -0,0 +1,166 @@
+import type { TextDocumentContentChangeEvent } from 'vscode-languageserver-protocol';
+
+export interface DelegatingFileSystemProviderOptions {
+    toResource(resource: URI): URI;
```
added examples

akosyakov commented on packages/filesystem/src/browser/file-service.ts:
```diff
@@ -0,0 +1,1505 @@
+
+    protected async getEncodingForResource(resource: URI, preferredEncoding?: string): Promise<string> {
+        let provider = await this.withProvider(resource);
+        while (provider instanceof DelegatingFileSystemProvider) {
```
replaced with `DelegatingFileSystemProvider.toUnderlyingResource`

akosyakov commented on packages/filesystem/src/common/io.ts:
```diff
@@ -0,0 +1,111 @@
+import { CancellationToken, cancelled as canceled } from '@theia/core/lib/common/cancellation';
+import { FileSystemProviderWithOpenReadWriteCloseCapability, FileReadStreamOptions, ensureFileSystemProviderError } from './files';
+
+export interface ICreateReadStreamOptions extends FileReadStreamOptions {
```
overlooked it, should be fixed

akosyakov commented on packages/filesystem/src/common/remote-file-system-provider.ts:
```diff
@@ -0,0 +1,399 @@
+}
+
+@injectable()
+export class RemoteFileSystemProvider implements Required<FileSystemProvider>, Disposable {
```
To make that they don't use any node or browser APIs and can be reused by downstreams in other contexts, i.e running some fs provider in the forked process and use both classes on the backend for forwarding and so on.

akosyakov commented on packages/filesystem/src/common/remote-file-system-provider.ts:
```diff
@@ -0,0 +1,399 @@
+}
+
+@injectable()
+export class FileSystemProviderServer implements RemoteFileSystemServer {
```
added js-doc

akosyakov commented on packages/mini-browser/src/browser/mini-browser-content.ts:
```diff
@@ -254,18 +251,18 @@ export class MiniBrowserContent extends BaseWidget {
+            if (await this.fileService.exists(new URI(location))) {
                 const fileUri = new URI(location);
-                const watcher = await this.fileSystemWatcher.watchFileChanges(fileUri);
+                const watcher = this.fileService.watch(fileUri);
```
yes, watch API is sync now

akosyakov commented on packages/mini-browser/src/browser/mini-browser-content.ts:
```diff
@@ -254,18 +251,18 @@ export class MiniBrowserContent extends BaseWidget {
 
     protected async listenOnContentChange(location: string): Promise<void> {
         if (location.startsWith('file://')) {
```
really good catch!

akosyakov commented on packages/monaco/src/browser/monaco-editor-model.ts:
```diff
@@ -68,16 +71,19 @@ export class MonacoEditorModel implements ITextEditorModel, TextEditorDocument {
+    readonly onDidChangeEncoding = this.onDidChangeEncodingEmitter.event;
+
+    private preferredEncoding: string | undefined;
+    private _contentEncoding: string | undefined;
```
fixed

akosyakov commented on packages/monaco/src/browser/monaco-editor-model.ts:
```diff
@@ -42,6 +42,9 @@ export interface MonacoModelContentChangedEvent {
     readonly contentChanges: TextDocumentContentChangeEvent[];
 }
 
+export interface MonacoEditorModelOptions {
```
removed

akosyakov commented on packages/monaco/src/browser/monaco-snippet-suggest-provider.ts:
```diff
@@ -138,14 +139,15 @@ export class MonacoSnippetSuggestProvider implements monaco.languages.Completion
             toDispose.push(this.fromJSON(snippets, options));
         } catch (e) {
-            if (!FileSystemError.FileNotFound.is(e) && !FileSystemError.FileIsDirectory.is(e)) {
+            if (!(e instanceof FileOperationError)) {
```
unfortunately I could not find any comment in commit or PR history why it was added, my suspicious are that there were  bad VS Code extensions or something like that. We could remove it all together and try if you want, but I don't think that PR is really changing semantic.

tsmaeder replied:
Fine by me under the principle of "it's not worse than before" ;-)

akosyakov commented on packages/navigator/src/browser/navigator-diff.ts:
```diff
@@ -71,11 +72,12 @@ export class NavigatorDiff {
+            const stat = await this.fileService.resolve(uri);
+            return stat.isDirectory;
+        } catch (e) {
+            if (e instanceof FileOperationError && e.fileOperationResult === FileOperationResult.FILE_NOT_FOUND) {
```
to align with semantic on master: https://github.com/eclipse-theia/theia/blob/fd239889618f16b8ac48d50a754bf363ee5230c9/packages/navigator/src/browser/navigator-diff.ts#L75-L76

`!stat` - means file not found

probably method should be renamed to `isFile` and return `false` always on exception, but i would rather leave such refactoring to a separate PR

akosyakov commented on packages/navigator/src/browser/navigator-widget.tsx:
```diff
@@ -175,11 +156,14 @@ export class FileNavigatorWidget extends FileTreeWidget {
-                if (this.model.copy(uri)) {
-                    event.preventDefault();
-                }
+                event.preventDefault();
```
it is no-op for second call and we should call it if at least one URI was moved

akosyakov commented on packages/plugin-ext-vscode/src/browser/plugin-vscode-contribution.ts:
```diff
@@ -0,0 +1,45 @@
+import { DelegatingFileSystemProvider } from '@theia/filesystem/lib/common/delegating-file-system-provider';
+
+@injectable()
+export class PluginVSCodeContribution implements FileServiceContribution {
```
We usually have one contribution from an extension which implements many contribution, i.e. `NavigatorContribution` implement `CommandContribution`, `MenuContribution` and so on. The same is here `PluginVSCodeContribution` can implement any additional application contributions.

tsmaeder replied:
Fair enough.

akosyakov commented on packages/plugin-ext-vscode/src/browser/plugin-vscode-contribution.ts:
```diff
@@ -0,0 +1,45 @@
+export class PluginVSCodeContribution implements FileServiceContribution {
+
+    registerFileSystemProviders(service: FileService): void {
+        this.mapSchemas(service, Schemes.vscodeRemote, 'file');
```
It is scheme to resolve resource in a workspace. By default it assumes to be next on the same machine where Theia backend is running, i.e. `file` scheme. One could rebind `PluginVSCodeContribution` to delegate `vscode-remote` to some other fs provider if it is required. Some VS Code extensions could create such URIs to open documents and so on.

tsmaeder replied:
Is there documentation describing this for VS Code somewhere? A link would be helpful understanding what's happening here.

akosyakov replied:
I could not find any docs and it is only my guess from VS Code source code. I am not really sure whether we need it at all, Maybe it is just internal  VS Code schema which is used to look up resources when the extension host process is running on remote machine and we don't really need it. 🤷‍♂️ Although i've seen it is used in built-in extensions: https://github.com/microsoft/vscode/blob/f74e473238aca7b79c08be761d99a0232838ca4c/extensions/git/src/timelineProvider.ts#L77

tsmaeder replied:
:sigh:

akosyakov commented on packages/plugin-ext/src/main/browser/editor/untitled-resource.ts:
```diff
@@ -101,8 +101,8 @@ export class UntitledResource implements Resource {
-    async saveContentChanges(changes: TextDocumentContentChangeEvent[], options?: { encoding?: string, overwriteEncoding?: string }): Promise<void> {
-        if (!this.fileResource) {
+    async saveContentChanges(changes: TextDocumentContentChangeEvent[], options?: ResourceSaveOptions): Promise<void> {
+        if (!this.fileResource || !this.fileResource.saveContentChanges) {
```
What do you mean by the second case?

The contract is that  if `saveContentChanges` is defined then it is supported if not then it is not. It is the same on master. I like an idea of capabilities, but it will be another refactoring.

tsmaeder replied:
I guess that it depends on the underlying file system whether "saveContentChanges" is implemented. So the resource can't really decide that.

akosyakov commented on packages/monaco/src/browser/monaco-editor-model.ts:
```diff
@@ -100,23 +105,44 @@ export class MonacoEditorModel implements ITextEditorModel, TextEditorDocument {
     async saveWithEncoding(encoding: string): Promise<void> {
-        return this.scheduleSave(TextDocumentSaveReason.Manual, this.cancelSave(), encoding)
-            .then(() => { this.preferredEncoding = encoding; });
+        if (!this.setPreferredEncoding(encoding)) {
```
> So if the encoding is not different from what we have, we do not save the file? Can it be dirty at this point?

It means that the document was already loaded in such encoding, file was already saved or save for such encoding is already scheduled.

tsmaeder replied:
Hmhh...so even if the file was loaded with the same encoding...if it's dirty, shouldn't we call save anyway?

akosyakov replied:
You mean that method is called `save`, but it does not save always? Would it remove confusion if I rename it to `setEncoding`?

akosyakov replied:
I've replaced `reopenWithEncoding` and `saveWithEncoding` with `setEncoding`. It also aligns well with `getEncoding`.

tsmaeder replied:
Not really: I still don't understand why it makes sense to start with a dirty file and save it if the encoding changed and not save it if the encoding is the same. My suspicion is that this case does not happen, but that is not obvious from the code.

akosyakov replied:
> I still don't understand why it makes sense to start with a dirty file and save it if the encoding changed and not save it if the encoding is the same.

The only client is the status bar entry allowing a user to save file with encoding. Right now behaviour is aligned with VS Code, i.e. the same encoding is no-op, no side effects like saving dirty changes. We could save always I don't see the issue with it, especially when a user command is called `Save with encoding...`.

> My suspicion is that this case does not happen, but that is not obvious from the code.

I don't expect users to save with the current encoding either. I guess they simply never select the same encoding or close the quick pick.

tsmaeder replied:
Ok.

akosyakov commented on packages/monaco/src/browser/monaco-editor-model.ts:
```diff
@@ -422,6 +449,7 @@ export class MonacoEditorModel implements ITextEditorModel, TextEditorDocument {
             await Resource.save(this.resource, { changes, content, options: { encoding, overwriteEncoding, version } }, token);
             this.contentChanges.splice(0, changes.length);
             this.resourceVersion = this.resource.version;
+            this.updateContentEncoding();
```
Set a preferred encoding is always called with overwrite true, so it could happen only if there is then a concurrent save which cancels overwrite save. It was also bothering me on master, but I decided don't change it since did not want to refactor much here. I will file a follow-up issue. I don't think it is caused by this PR.

akosyakov commented:
@tsmaeder I've addressed your comments. Could you have a look please?

@eclipse-theia/ecd-theia-committers CQ is approved, I would like to merge it after next release, i.e. this Thursday if no objections

tsmaeder commented:
@akosyakov all my comments are resolved.

yunuscukran commented:
Does this update change https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#uripath?

akosyakov commented:
@yunuscukran updated guidelines
